Initial implementation for transaction export

boxfoot opened the pull request:
I know this isn't on the Trello board, but it's a particular need of mine (for tax prep, since there's no reporting function yet) so I thought I would give it a stab as a way to learn the toolkit.

I definitely welcome feedback on implementation, user experience, code formatting, etc.

To export your transactions, go to the "All Accounts" screen.  There should be a new "Export" button in the header.  Click it.

Niictar commented:
@boxfoot

This is cool. When I test this, I need a bit of extra work to get the export button to appear. However, once I got it to appear, the actual download seems to work well. I wonder if we'll be able to use some of the logic in your new feature to handle some other interesting behaviours with the accounts view.

In order to get the download button to appear, I had to run ynabToolKit.exportTransactions.invoke() manually from the console. Everything else worked. So, the trick is to reliably run invoke() and we should be set.

Right now, the part of the code that doesn't quite work looks like this:

```javascript
      // ####################
      // # FEATURE CODE 
      // ####################

      this.invoke = function() {

        // Build the UI button
        var exportButton = '<button id="toggleSplits" class="ember-view button"><i class="ember-view flaticon stroke download-document"></i> Export </button>';
        this.button = $(exportButton).on('click', downloadTransactions);
        
        // Determine whether to show the button on this screen
        this.observe();
      };
      

      this.observe = function() {
        // Only show export button on the 'All Accounts' screen      
        if (location.href.endsWith('/accounts')) {
          this.button.insertAfter(".accounts-toolbar .accounts-toolbar-edit-transaction");
        } else {
          this.button.detach();
        }
          
      };
    };

  } else { //... line 179
```

Presently, this.invoke() calls this.observe(). What we actually want to do is use this.observe() to call this.invoke() when necessary. Another file which @blargity is currently setting up to build automatically is feedChanges.js. This file, which you'll need to update manually for now, contains a list of try...catch statements that is called when MutationObserver finds changes on the page. If you add the following below line 59 in feedChanges.js:

```javascript
try {
        ynabToolKit.exportTransactions.observe(changedNodes);
      } catch(err) {/* ignore */}
```

Your function will get called whenever there is a change observed on the page. Then we just setup code that evaluates the situation and decides if we need to run this.invoke().

I have refactored the relevant parts of the code below, which causes the export button to appear reliably (with that update to feedChanges.js):

```javascript
      // ####################
      // # FEATURE CODE
      // ####################

      this.invoke = function() {

        var exportButton = '<button id="export-transactions" class="ember-view button"><i class="ember-view flaticon stroke download-document"></i> Export </button>'
        $(exportButton).insertAfter(".accounts-toolbar .accounts-toolbar-edit-transaction");
        $('#export-transactions').click(function() { downloadTransactions(); });

      },

      this.observe = function(changedNodes) {

        // Only show export button on the 'All Accounts' screen
        if (location.href.endsWith('/accounts') && !$('.download-document').length ) {
          ynabToolKit.exportTransactions.invoke();
        }

      };

    };
    ynabToolKit.exportTransactions.invoke(); // run once on page load

  } else { // line 175 ...
```

Note in this case we aren't checking the contents of changedNodes object at all, since we are using the URL instead. However, you may want to use it in the future:

````javascript
if ( changedNodes.has('important-class') { // invoke function }
```

I hope this helps!

boxfoot commented:
Thanks for the feedback - I'll take a closer look over the weekend.

Niictar commented:
@boxfoot I did notice in my example that I introduced a regression by not properly removing the button if the user navigates away from the All Accounts page. For example, if they click on All Accounts and then to their Checking, the export all button persists. So, you'll want to allow watch out for that.

boxfoot commented:
Thanks - I actually saw that, too.  But otherwise your edits make a ton of sense.  I was confused by how `invoke` and `observe` work, which seems kind of magical.  Is that documented anywhere yet?  I know it's a work in progress...

BTW, just commented about this on Trello, too: https://trello.com/c/mEPvAQUE

Niictar commented:
> which seems kind of magical. Is that documented anywhere yet?

Yea, I wish 'kind of magical' was a positive thing, but I agree with you :sweat_smile: . Once the dust settles (a bunch of this was only recently fully refactored), documentation is next on the TODO list.

blargity commented:
Yup, plan is to document both settings.json files and the Act on Change stuff once Python scripting of it is done.

It'll still be a bit magical, but hopefully it'll be a documented "this is what's behind the curtain" kind of magical.

Thanks for the pull request! Let me know when you're ready for us to have another look at it. :+1:

boxfoot commented:
OK, ready for review again.  I think this addresses all the issues raised (although please check my use of invoke & observe).

Also - today my Chrome kept crashing when I tried to load the toolkit in Developer Mode (pretty sure that's not just about my changes) so I wasn't actually able to test this myself, hopefully I haven't missed anything obvious as a result.  Feel free to ignore if you want to wait until I have a chance to test properly.

blargity commented:
Howdy!

We really need to document this and quickly, so sorry! Observe gets called any time any DOM node changes on the page. You definitely want to look for changes you care about and then do stuff, rather than calling invoke for every change that comes your way. I suspect if you comment out your invoke() call in the observe method Chrome might stop crashing on you.

In your case I'd use observe changes to figure out whether they're on the accounts page or not. I just did this with the new reports tab I'm building here: https://github.com/blargity/toolkit-for-ynab/blob/net-worth-report/src/common/res/features/reports/main.js

Main bit you'll care about is this:

```
        // This code is checking for when the user navigates away from the reports page. You'd want to check for when they leave the accounts page.

        this.observe = function(changedNodes) {
          if (changedNodes.has('navlink-budget') || changedNodes.has('navlink-accounts')) {

            // The nav divs have these classes, so we know we care if they change.
            if ($('.navlink-budget').hasClass('active') || $('.navlink-accounts').hasClass('active')) {

              // The user has left the reports page.
              // We're no longer the active page.
              $('.navlink-reports').removeClass('active');

              // And we should restore the elements we replaced earlier.
              $('#reportCanvas').remove();
              $('div.scroll-wrap').show();
            }
          }
        };
```

Hope that helps, and yup, we'll definitely put pen to paper on this very soon!

Niictar commented:
I'll add that if you set 'ynabToolKit.debugNodes = true' in the console, it will start printing out everything MutationObserver sees when the page changes.

blargity commented:
Also another team member just found ```if (changedNodes.has('budget-header-flexbox') { /* stuff */ }``` for when you're on the budget page. There might be something similar for accounts you can look for.

blargity commented:
This feature's on the Trello board now (https://trello.com/c/nPPUqP71/183-transaction-export-to-csv-format), just let me know if you want to be added to the Trello team or if you only plan to contribute this feature. :+1:

blargity commented:
I took a first pass at the docs. Hope they help and let me know if there's anything that we can improve:

https://github.com/blargity/toolkit-for-ynab/blob/master/src/common/res/features/HOW_TO_BUILD_FEATURES.md

boxfoot commented:
@blargity the docs are really helpful -- I think I now understand how changedNodes works (it's different than what I thought) and get how I was crashing the app before.  This PR also includes some suggested updates to the docs to make it even clearer.

FWIW, I found `ynabToolKit.debugNodes = true` to be an overwhelming amount of mostly unhelpful detail.  Since our features can only consume the final set of classnames anyway, I think it would be great if we could have a debug or developer flag which just outputs that to the console.  (That's how I did my development -- while looking for good hooks, I just did `this.observe = function (changedNodes) { console.log(changedNodes); };` and it was much more digestible).  Would you be open to a PR for that?

With all that help, I think this is finally ready for review again!

And sure, add me to the Trello board.  I'm enjoying this, not sure how much time I'll have to contribute but I can try.

blargity commented:
Yup, absolutely open to a PR for a debug option that only outputs the classes that are actually getting passed to you.

I'll add you to the Trello board, what's your username over there?

I'll have a look at this now.

blargity commented:
This is awesome! The only suggestion I'd have is it might be better to replace "Internal Master Category", "Immediate Income SubCategory", "Split SubCategory" with names that are a bit more user friendly, but this is totally ready for release as-is.

Thank you.

blargity commented:
And your feature is now released! :+1:

boxfoot commented:
Wow - very cool to see this live.  Thanks for the support!

Great suggestion, BTW - I just created #158 so that it won't get lost before I have a chance to work on it.  (Thought -- maybe we could label issues with the feature they're related to e.g. `feature-export-transactions`?)
